### PR TITLE
 fix: Use name instead of UID for pivot column download [DHIS2-9026]

### PR DESCRIPTION
**Pull request by maikelarabori**
Creating a simple object (DimensionDescriptor.java) to map raw dimension identifiers and its related type so it can be used in a later stage when setting the report headers during the download in the Pivot table app.

The main logic relies in the method:

`DimensionDescriptor.getDimensionIdentifierFor( final String dimension,
        final List<DimensionDescriptor> dimensionDescriptors )`

Basically, based on the given dimension and the given DimensionDescriptor list, this method will retrieve the respective dimension identifier. See the examples below.

- For regular dimensions:
a "dimension" `dx` will have a type of `DimensionType.DATA_X`. Hence the dimension identifier returned will be `dx`.

- For dynamic dimensions:
a "dimension" `mq4jAnN6fg3` (of an org unit, for example), will have a type of `DimensionType.ORGANISATION_UNIT`. Hence the dimension identifier returned by this method will be `ou`. So we are basically mapping `mq4jAnN6fg3` to `ou`.

It will allow the column's header, in the download file, to follow the same standard as the regular dimensions. In this case, this will result in: _**organisationunitname**_

Without this fix, the column header for this example would be: _**mq4jAnN6fg3name**_

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java [hunk 1]
```diff
@@ -1025,6 +1028,30 @@ public void setYearlySeries( List<String> yearlySeries )
...
+
+    /**
+     * This method will hold the mapping of a dimension and its respective formal
+     * type.
+     *
+     * @param dimension: the dimension, which should also be found in
+     *        "{@link #columnDimensions}" and "{@link #rowDimensions}".
+     * @param dimensionType: the formal dimension type. See {@link DimensionType}
+     */
+    public void holdDimensionDescriptor( final String dimension, final DimensionType dimensionType )
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java:1050]: Maybe rename to `addDimensionDescriptor` to make it similar to other naming in the system.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java [hunk 2]
```diff
@@ -342,6 +343,8 @@
 
     private transient List<List<DimensionalItemObject>> gridRows = new ArrayList<>();
 
+    private transient List<DimensionDescriptor> dimensionDescriptors = new ArrayList<>( 0 );
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java:346]: Remove "0" here.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/DimensionDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                        ? dimensionDescriptor.getDescriptiveValue()
+                        : defaultIfEmpty( metaData.get( valueToRetrieve ), valueToRetrieve );
+                }
+            }
+        }
+
+        return descriptiveValue;
+    }
+
+    private String getDescriptiveValue()
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/DimensionDescriptor.java:86]: Do you think you could add a description about what is going here in the PR and in Javadoc?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java (hunk 1 above)
Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java:1050]: Done

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java (hunk 2 above)
Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/Visualization.java:346]: Done

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/DimensionDescriptor.java (hunk 3 above)
Comment (maikelarabori) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/visualization/DimensionDescriptor.java:86]: Done

**Comment by larshelge** [comment]
Nice work here @maikelarabori . There is a conflict in this PR now, I think because of the recent refactor, which must be fixed. Otherwise I think it is ready to be merged.

**Review by larshelge — APPROVED** [review]

**Pull request merged**